src/Sylius/Component/Core/StateGuard/CompleteStepGuard.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Core\StateGuard;
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Order\Requirements\RequiredBillingAddressSpecification;
+use Sylius\Component\Core\Order\Requirements\RequiredNonEmptyCartSpecification;
+use Sylius\Component\Core\Order\Requirements\RequiredPaymentSpecification;
+use Sylius\Component\Core\Order\Requirements\RequiredShippingSpecification;
+
+class CompleteStepGuard implements OrderGuardInterface
+{
+    public function isSatisfiedBy(OrderInterface $order): bool
+    {
+        $requirements = new RequiredNonEmptyCartSpecification();
+        $requirements
+            ->and(new RequiredBillingAddressSpecification())
+            ->and(new RequiredShippingSpecification())
+            ->and(new RequiredPaymentSpecification())
```

Comment:
Do we really need so many levels of abstractions?
- imo logic from this `specifications` can by directly in guard (they are 4 lines services)
- this guard should use the rest of the guard, it makes this code much easier to read.
- we should; avoid creating classes by `new` why don't you inject it like normal services?
- and definitely, we are missing scenarios, each service returns true or false so we have 8 situations but only one scenario (the similar situation for the rest of the services)
- it is checkout, so we should (MUST) test it as good as it is possible because if sth does not work it always is code without tests :)

but, i really like rost on my person so WDYT @lchrusciel @GSadee @Zales0123

Reply:
Yup, we can discuss that. The Specification provides reusable business logic (Domain). Guard classes are basically the Framework or Infrastructure layer of the checkout system. I could create even more abstraction and provide Strategy over coupled Specification or not use Guards at all. The less neat part is that they are all in the Core namespace, but if I will move it somewhere else they will be coupling between components or bundles. Define best practices as I'm not aware of Sylius's approach :)


PS. We don't want to provide specifications via DI - it's gonna look terrible.

Reply:
I wonder about the way of customizing that by the end application, if someone would like to change the specification, there is a need to define a new one and overwrite the proper guard. Probably it is simple enough 🤔 
I know that I'm not helpful enough, the concept looks nice, easy to adjust and probably quite easy to customize, but it is also a new one in Sylius, there will be a need for proper documentation of it.